File: 이준영/백준/Gold/20056. 마법사 상어와 파이어볼/마법사 상어와 파이어볼.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.StringTokenizer;

public class Main {
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		int[] dy = { -1, -1, 0, 1, 1, 1, 0, -1 };
		int[] dx = { 0, 1, 1, 1, 0, -1, -1, -1 };

		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());
		int K = Integer.parseInt(st.nextToken());
		Map<Integer, ArrayList<int[]>> map = new HashMap<>();

		for (int i = 0; i < M; i++) {

			st = new StringTokenizer(br.readLine());
			int y = Integer.parseInt(st.nextToken());
			int x = Integer.parseInt(st.nextToken());
			int m = Integer.parseInt(st.nextToken());
			int s = Integer.parseInt(st.nextToken());
			int d = Integer.parseInt(st.nextToken());
			ArrayList<int[]> lst = new ArrayList<>();
			lst.add(new int[] { m, s, d });
			map.put((y - 1) * 50 + (x - 1), lst);

		}
		while (K > 0) {
//			for (int i : map.keySet()) {
//				ArrayList<int[]> lst2 = map.get(i);
//				for(int[] tmp:lst2) {
//					System.out.println(i/50+" "+i%50+" "+tmp[0]+" "+tmp[1]+" "+tmp[2]);
//				}
//			}
//			System.out.println();
			Map<Integer, ArrayList<int[]>> map2 = new HashMap<>();
			for (int i : map.keySet()) {
				ArrayList<int[]> lst2 = map.get(i);
				for (int[] tmp2 : lst2) {
					int y = i / 50;
					int x = i % 50;
					int speed = tmp2[1] % N; // 속도를 N으로 나눈 나머지만 사용
					int ny = (y + speed * dy[tmp2[2]]) % N;
					int nx = (x + speed * dx[tmp2[2]]) % N;
					ny = (ny + N) % N;
					nx = (nx + N) % N;
					if (map2.containsKey(ny * 50 + nx)) {
						ArrayList<int[]> lst = map2.get(ny * 50 + nx);
						lst.add(tmp2);
						map2.put(ny * 50 + nx, lst);
					} else {
						ArrayList<int[]> lst = new ArrayList<>();
						lst.add(tmp2);
						map2.put(ny * 50 + nx, lst);
					}
				}
			}

//			for (int i : map2.keySet()) {
//				ArrayList<int[]> lst2 = map2.get(i);
//				for(int[] tmp:lst2) {
//				System.out.println(i/50+" "+i%50+" "+tmp[0]+" "+tmp[1]+" "+tmp[2]);
//				}
//			}
//			System.out.println();

			map.clear();
			for (int tmp : map2.keySet()) {
				ArrayList<int[]> arr = map2.get(tmp);
				if (arr.size() == 1) {
					map.put(tmp, arr);
					continue;
				}
				int count = 0;
				int sum = 0;
				int check = 0;
				int v = 0;
				int d = 1;
				for (int[] lst : arr) {
					count += 1;
					sum += lst[0];
					v += lst[1];
					if (lst[2] % 2 == 0) {
						check += 1;
					}
				}
				if (sum < 5) {
					continue;
				}
				sum = sum / 5;
				v = v / count;
				if (check == count || check == 0) {
					d = 0;
				}
				ArrayList<int[]> lst = new ArrayList<>();
				for (int i = 0; i < 4; i++) {
					lst.add(new int[] { sum, v, 2 * i + d });
				}
				map.put(tmp, lst);
			}
			K = K - 1;
		}
		int count = 0;
		for (int i : map.keySet()) {
			ArrayList<int[]> lst2 = map.get(i);
			for (int[] tmp : lst2) {
				count = count + tmp[0];
			}
		}
		System.out.println(count);

	}
}

File: 이준영/백준/Gold/20056. 마법사 상어와 파이어볼/마법사 상어와 파이어볼.py
n, m, k = map(int, input().split())

dy = [-1, -1, 0, 1, 1, 1, 0, -1]
dx = [0, 1, 1, 1, 0, -1, -1, -1]

lst = [0] * m
for i in range(m):
    x, y, m, s, d = map(int, input().split())
    lst[i] = [x - 1, y - 1, m, s, d]

for _ in range(k):
    arr = [[[] for _ in range(n)] for _ in range(n)]
    tmp_lst = []
    for y, x, m, s, d in lst:
        y = (y + s * dy[d]) % n
        x = (x + s * dx[d]) % n
        arr[y][x].append((m, s, d))

    for i in range(n):
        for j in range(n):
            if len(arr[i][j]) == 0:
                continue
            elif len(arr[i][j]) == 1:
                tmp = arr[i][j][0]
                tmp_lst.append([i, j, tmp[0], tmp[1], tmp[2]])
            else:
                cnt = 0
                even = 0
                speed = 0
                weight = 0
                for m, s, d in arr[i][j]:
                    weight += m
                    speed += s
                    cnt += 1
                    if d % 2 == 0:
                        even += 1

                weight = weight // 5
                speed = speed // cnt
                if even == cnt or even == 0:
                    start = 0
                else:
                    start = 1

                if weight == 0:
                    continue
                else:
                    for k in range(4):
                        tmp_lst.append([i, j, weight, speed, k * 2 + start])

    lst = tmp_lst

tot = 0
for y, x, m, s, d in lst:
    tot += m
print(tot)
